projects/m5/003-probability-calculator/js/src/prob_calculator.js

```diff
@@ -0,0 +1,121 @@
+// @ts-check
+/**
+ * @fileoverview Create a Probability Calculator.
+ * @author Salvatore Quagliariello
+ * @module prob_calculator
+ */
+
+/**
+ * @classdesc A Class to create Hat Objects.
+ */
+class Hat {
+    /**
+     * Accept an undefined number of strings that will be converted into object
+     * with color as key and number as value. The objects will then be pushed
+     * into the `contents` Array.
+     * @constructor
+     * @param  {...string} balls A string describing a color the number of balls (e.g blue=2).
+     */
+    constructor(...balls) {
+        /** @private @constant {Array<Object>} */
+       this.contents = [];
+
+       balls.forEach(item => { 
+        const color = item.split(`=`)[0];
+        const numberOfBalls = parseInt(item.split(`=`)[1]);
+
+            for (let i = 0; i < numberOfBalls; i++) {
+                this.contents.push(color);
+            };
+       });
+    }
+
+    /**
+     * Takes a number as argument. The method mimin a draw from an Hat Object.
+     * The draw is completely random as it is based on Fisher-Yates algorithm.
+     * It shuffle the `contents` Array of the Object and pick the first entries,
+     * to push it to another Array, and delete it from the `contents` Array.
+     * If the Number used as argument is greater than the Array size, the 
+     * nethod returns the whole array.
+     * @param {number} numberOfBalls The number of balls to draw from the Hat Object.
+     * @returns {Array<string>}
+     */
+    draw(numberOfBalls) {
+        const balls = [...this.contents];
+        const pickedBalls = [];
+
+        if(numberOfBalls >= this.contents.length) {
+            this.contents = [];
+            return balls;
+        };
+
+        for (let k = 0; k < numberOfBalls; k++) {
+            for (let i = balls.length - 1; i > 0; i--) {
+              const j = Math.floor(Math.random() * (i + 1));
+              const temp = balls[i];
+              balls[i] = balls[j];
+              balls[j] = temp;
+            };
+            this.contents.splice(this.contents.indexOf(balls[0]), 1);
+            balls.splice(0, 1);
+            pickedBalls.push(balls[0]);
+        };
```

Comment:
Optional Tip:
`const random = Math.floor(Math.random() * months.length);`
`pickedBalls.push(this.contents[random]`
`this.contents.splice(random, 1)`

If you want you can try something like this ;-)